exist-core/src/main/java/org/exist/xquery/functions/fn/transform/Transform.java

```diff
@@ -0,0 +1,485 @@
+import java.net.URI;
+import java.time.LocalDateTime;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.evolvedbinary.j8fu.tuple.Tuple.Tuple;
+
+/**
+ * Implementation of fn:transform.
+ *
+ * This is the core of the eval() function of fn:transform
+ * It is a separate class due to the multiplicity of options that must be dealt with,
+ * which lead to too much code in a single file.
+ *
+ * This class contains the core of the logic
+ * - create an XSLT compiler (if we don't have one compiled already, which matches our stylesheet)
+ * - set parameters on the compiler
+ * - create a transformer from the compiler
+ * - set parameters on the transformer (inputs)
+ * - invoke the transformation
+ * - deliver and postprocess the output in the required form
+ *
+ * The parsing and checking of the options to fn:transform is isolated in {@link Options}
+ * Delivery and output of the result, depending on the required format, is in {@link Delivery}
+ *
+ * @author <a href="mailto:adam@evolvedbinary.com">Adam Retter</a>
+ * @author <a href="mailto:alan@evolvedbinary.com">Alan Paxton</a>
+ */
+public class Transform {
+
+    private static final Logger LOGGER =  LogManager.getLogger(org.exist.xquery.functions.fn.transform.Transform.class);
+    private static final org.exist.xquery.functions.fn.transform.Transform.ErrorListenerLog4jAdapter ERROR_LISTENER = new Transform.ErrorListenerLog4jAdapter(Transform.LOGGER);
+
+    //TODO(AR) if you want Saxon-EE features we need to set those in the Configuration
+    static final Configuration SAXON_CONFIGURATION = new Configuration();
+    private static final Processor SAXON_PROCESSOR = new Processor(org.exist.xquery.functions.fn.transform.Transform.SAXON_CONFIGURATION);
+
+    static final Convert.ToSaxon toSaxon = new Convert.ToSaxon() {
+        @Override
+        DocumentBuilder newDocumentBuilder() {
+            return SAXON_PROCESSOR.newDocumentBuilder();
+        }
+    };
+
+    private static final Cache<String, XsltExecutable> XSLT_EXECUTABLE_CACHE = Caffeine.newBuilder()
+            .maximumSize(25)
+            .weakValues()
+            .build();
+
+    private final XQueryContext context;
+    private final FnTransform fnTransform;
+
+    public Transform(final XQueryContext context, final FnTransform fnTransform) {
+        this.context = context;
+        this.fnTransform = fnTransform;
+    }
+
+    public Sequence eval(final Sequence[] args, final Sequence contextSequence) throws XPathException {
+
+        final Options options = new Options(context, fnTransform, (MapType) args[0].itemAt(0));
+
+        //TODO(AR) Saxon recommends to use a <code>StreamSource</code> or <code>SAXSource</code> instead of DOMSource for performance
+        final Optional<Source> sourceNode = Transform.getSourceNode(options.sourceNode, context.getBaseURI());
+
+        if (options.xsltVersion == 1.0f || options.xsltVersion == 2.0f || options.xsltVersion == 3.0f) {
```

Comment:
surprised that this works, as 3 could be 2.99999999 hence the comparison fails. Should use Flood.equals() instead?

Reply:
We're probably OK if the version is never subjected to arithmetic ? Anyway, I got paranoid and built an XSLTVersion class with major/minor which reads from a BigDecimal.